Fix drawImage regression with particular use case of 9 arguments.

asturur opened the pull request:
So this should fix 95% of the cases, and almost all real usage cases.

I think my code introduced another regression that i have no time to verify and i ll verify ( and in case fix ) as soon as possible, is NOT visible from the screenshots, when a scaling is applied on the canvas, and is not considered in drawImage code with the extra canvas, some detail loss will happen. I have an idea how to fix this, but i think should be a different PR.

This will happen only when a small destination size is specified but with upscaled receiving context.
Is a weird use case that will likely happen on higher level libraries like fabricJS and not on direct canvas usage ( since the pattern is a bit strange )

I changed a bit the tests, i should reverify them with the 1.6.9 to be sure the failing one are failing for my code or also before.
I set the globalAlpha of the globalCompositeOperation tests lower than 1 so that particular interaction can be noted.

## NEW BUG UNRELATED WITH THIS FIX

something is wrong with the copy version
![image](https://user-images.githubusercontent.com/1194048/45925155-751cbe00-bf10-11e8-91bf-225e6e31ce63.png)

saturation is not perfect yet
![image](https://user-images.githubusercontent.com/1194048/45925163-a1d0d580-bf10-11e8-8bdb-1ffa3a8ecf39.png)

## FIXES RELATED TO THIS PR
<details>

![image](https://user-images.githubusercontent.com/1194048/45925178-bd3be080-bf10-11e8-8fc7-8be1d91c140c.png)
![image](https://user-images.githubusercontent.com/1194048/45925181-d0e74700-bf10-11e8-86c6-5993ed913cc3.png)
![image](https://user-images.githubusercontent.com/1194048/45925185-de9ccc80-bf10-11e8-8429-e808adcb0d95.png)
![image](https://user-images.githubusercontent.com/1194048/45925187-ea888e80-bf10-11e8-87ea-38bcaf1efdf8.png)
![image](https://user-images.githubusercontent.com/1194048/45925192-f3796000-bf10-11e8-821b-4d7ad73191c7.png)
![image](https://user-images.githubusercontent.com/1194048/45925196-fd9b5e80-bf10-11e8-9888-89bfc5363519.png)
![image](https://user-images.githubusercontent.com/1194048/45925202-09872080-bf11-11e8-932d-dc5af6a81a2e.png)
![image](https://user-images.githubusercontent.com/1194048/45925206-14da4c00-bf11-11e8-9a5d-29bed8fe8d11.png)
![image](https://user-images.githubusercontent.com/1194048/45925210-1d328700-bf11-11e8-9af2-5c332e421d39.png)

</details>

## TESTS RESULT WITH MASTER CODE

<details>

![image](https://user-images.githubusercontent.com/1194048/45925237-b95c8e00-bf11-11e8-804d-2fab7e6b4427.png)
![image](https://user-images.githubusercontent.com/1194048/45925239-c8dbd700-bf11-11e8-88d7-d924e8a90e78.png)
![image](https://user-images.githubusercontent.com/1194048/45925242-d1341200-bf11-11e8-8e5d-80421b5f381e.png)
![image](https://user-images.githubusercontent.com/1194048/45925243-d98c4d00-bf11-11e8-9399-6519ea298898.png)
.... and so on ....

</details>

## IMAGE SAMPLING TEST
This test is the reason why we need to use an extend in first place and works fine.
![image](https://user-images.githubusercontent.com/1194048/45925254-f9237580-bf11-11e8-82d2-1e63940d8318.png)

<sub>zbjornson edited to put all the images in `<details>`</sub>

zbjornson commented:
> Do you feel like we should add the particular test case that discovered the bug?

Yeah, or make a small repro case similar to one of the existing drawImage tests -- I don't think we need another test image added to this repository to reproduce it.

asturur commented:
ok all add it.

otiai10 commented:
Hi, I'm trying to use `node-canvas` to mock HTML canvas element in my tests.

Since I'm facing a problem in which `drawImage` of `CanvasRenderingContext2D` can only accept **3 arguments** and rejects **9 arguments** call, like this

```
% jest

    TypeError: Expected object, number and number

      25 |     this.canvas.height = rect.size.h;
      26 |     const ctx: CanvasRenderingContext2D = this.canvas.getContext("2d");
    > 27 |     ctx.drawImage(
         |         ^
      28 |       this.img,
      29 |       rect.start.x, rect.start.y, rect.size.w, rect.size.h,
      30 |       0, 0, this.canvas.width, this.canvas.height,
      31 |     );
```

So, is this problem related to this issue, or should I create another specific issue post?

asturur commented:
Added a test similar to the issue:

With no fix:
![image](https://user-images.githubusercontent.com/1194048/46388558-35eb2b80-c6cd-11e8-8c86-d7417e3bad74.png)

with fix:
![image](https://user-images.githubusercontent.com/1194048/46388567-48fdfb80-c6cd-11e8-8fe9-78b20e4f0763.png)

asturur commented:
@LinusU @zbjornson  This should be done too.

asturur commented:
i ll check those cairo groups, i m not very deep into cairo

asturur commented on src/CanvasRenderingContext2d.cc:
```diff
@@ -1183,7 +1183,8 @@ NAN_METHOD(Context2d::DrawImage) {
   bool needScale = dw != sw || dh != sh;
-  bool needCut = sw != fw || sh != fh;
+  bool needCut = sw != source_h || sh != source_h || sx < 0 || sy < 0;
+  bool needClip = sx < 0 || sy < 0 || sw > source_w || sh > source_h;
```
yes those are both to address

asturur commented:
@zbjornson i changed needClip in needCairoClip, is the only thing i could do to differentiate clip and cut.
Also addressed the error on the var name.

Some of this clip and cut logic will be reworked in my next pr that should address the loss of details.
I try to figure out what cairo groups are on that PR and if they are applicable to do not create a full surface but keeping everything functional.

zbjornson commented:
Looks good, thanks!

Don't worry too much about groups. I think the only ("only") benefit they bring is a small, potential performance improvement, and possibly neater code. If you figure them out, great. :)

---

Travis got set to not build pull requests somehow. I just turned that back on, but I don't see a way to trigger a build for this PR. (The tests pass locally though.)

Will leave open for another ~day in case anyone else wants to look.

zbjornson commented:
Backported to v1.x in a0af58872fe3a713cb26d6dd3726871177d08cb3.

LinusU commented:
Released as:

- `canvas@2.0.0-alpha.18`
- `canvas@1.6.13`
